src/lib/util/db-lock.ts

```diff
@@ -0,0 +1,24 @@
+import { Client } from 'pg';
+import { IDBOption } from '../types';
+
+const defaultLockKey = 479341;
+
+export const withDbLock =
+    (dbConfig: IDBOption, timeout = 5000, lockKey = defaultLockKey) =>
+    <A extends any[], R>(fn: (...args: A) => Promise<R>) =>
+    async (...args: A): Promise<R> => {
+        const client = new Client({ ...dbConfig, query_timeout: timeout });
+        try {
+            await client.connect();
+            // wait to obtain a lock
+            await client.query('SELECT pg_advisory_lock($1)', [lockKey]);
```

Comment:
only one instance with a given key and DB name can run the SELECT. Other clients will await on the lock to free up. The timeout of 5000 prevents from infinite waiting. Realistically our migrations should run in under 5 seconds.